Add storyboards as a main menu background source

LeNitrous opened the pull request:
From #13299. Adds beatmap storyboards as a main menu background option.


https://user-images.githubusercontent.com/191335/121237125-acef0e00-c8d1-11eb-8d9c-ee98ca53ada3.mov

LeNitrous committed:
add beatmap with storyboard source

LeNitrous committed:
implement BeatmapBackgroundWithStoryboard

LeNitrous committed:
add new background type in BackgroundScreenDefault

LeNitrous committed:
fix inspect code issues and cleanup code

LeNitrous committed:
remove fade

peppy commented on osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs:
```diff
@@ -0,0 +1,34 @@
+                {
+                    AddInternal(loaded);
+                    if (Beatmap.Storyboard.ReplacesBackground)
+                        Sprite.FadeOut(300, Easing.OutQuint);
```
Is this matching how things are displayed during gameplay somehow? Curious why it's doing a fade as opposed to just showing the storyboard/video from the get go.

LeNitrous replied:
Backgrounds have scaling applied to account for parallax so it can't match gameplay 1:1.
For fading, I just checked player and there's no fading applied. I'll remove that in the next commit.

peppy replied:
Just keep in mind that with your current implementation, you're loading a background then displaying the storyboard on top. While this may visually work, I'm not sure if it's the best structure for this implementation (as opposed to not deriving from `BeatmapBackground` and handling whether the background or storyboard should be displayed locally, only initialising one or the other).

It's an isolated use case so wouldn't want to spend too much time on it (would probably be okay with how you have it until someone reports performance issues or otherwise), but if you're looking to put in the extra effort it may be worthwhile experimenting with changing the structure to as I propose here and see if it works well or not.

LeNitrous committed:
refactor BeatmapBackgroundWithStoryboard to reduce overhead

This avoids loading the sprite if its not needed and instead of hiding it, it is removed when the storyboard replaces the background or there is a video.

This also only initializes DrawableStoryboard if there are any elements in any layer.

peppy commented:
It's even worse now. You may have misunderstood my suggestion: it was to avoid inheriting from `BeatmapBackground` altogether, not adding hacky methods in the base class to work around it.

LeNitrous commented:
I tried to avoid duplicate code. Specifically the part around creating the sprite and setting its texture. But then again it seems that it should be fine enough to duplicate code in this scenario as there's little to copy to begin with?

peppy commented:
Might be best to just revert your last changes and go with it. It's going to take a bit more effort to do this in an elegant way, due to the fact you're providing your new class to a location which only handles `Background`, which comes with logic specific to displaying a single sprite.

I think it's probably fine to get that in and iterate from there based on whether people end up using this or not.

peppy committed:
Revert "refactor BeatmapBackgroundWithStoryboard to reduce overhead"

This reverts commit 277545bb0652c8e2e56c474293200b5b2c1755da.

peppy commented:
- Shouldn't play sound effects
- Currently shows all layers overlapping each other (test with [Don't stop me now](https://dev.ppy.sh/beatmapsets/5381#osu/26567))
- Shouldn't show a second time in `ScalingContainer` (see `ScalingBackgroundScreen`)

peppy reviewed (changes requested):
As mentioned

LeNitrous committed:
prevent scaling container from creating a storyboard background

LeNitrous committed:
apply suggestions

- Replace the sprite with a solid black box when a storyboard requests it.
- Create a new storyboard instance and exclude the fail layer as well as strip all samples from it
- Do not attempt in creating the storyboard when it isn't needed

LeNitrous committed:
use created storyboard to check for drawables instead

LeNitrous commented:
With Don't Stop Me Now, the overlapping sprites also happens in `Player` at version `2021.602.0`. I'm unsure if this issue is tracked or not.

peppy commented on osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs:
```diff
@@ -19,16 +22,24 @@ public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTe
+
+            foreach (var layer in storyboard.Layers)
+            {
+                if (layer.Name != "Fail")
```
these strings really need to be replaced with an enum...

peppy commented on osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs:
```diff
@@ -19,16 +22,24 @@ public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTe
+
+            if (storyboard.ReplacesBackground)
+            {
+                Sprite.Texture = Texture.WhitePixel;
```
Why did this logic change? The old way was fine.

LeNitrous replied:
When Screen Scaling is either Everything or Excluding Overlays, the scaling container background can be seen through
![image](https://user-images.githubusercontent.com/20495991/120601924-165bc080-c47d-11eb-931d-683eb329680c.png)

peppy replied:
So you made it black instead? I do not see that as a valid solution.

LeNitrous replied:
I've checked the behavior on `Player` itself and in the screenshot above is the intended. I'll just fade it out like before.

peppy commented on osu.Game/Graphics/Backgrounds/BeatmapBackgroundWithStoryboard.cs:
```diff
@@ -19,16 +22,24 @@ public BeatmapBackgroundWithStoryboard(WorkingBeatmap beatmap, string fallbackTe
-                    if (Beatmap.Storyboard.ReplacesBackground)
-                        Sprite.FadeOut(300, Easing.OutQuint);
-                });
+            var storyboard = new Storyboard { BeatmapInfo = Beatmap.BeatmapInfo };
```
This is wrong. You need to use the beatmap's storyboard. Things like `UseSkinSprites` are no longer set correctly.

LeNitrous replied:
Trying to add or remove elements from the working beatmap itself would also affect the storyboard when it is consumed later in gameplay. I can however pass in the `UseSkinSprites` value by:
```suggestion
            var storyboard = new Storyboard 
            {
                BeatmapInfo = Beatmap.BeatmapInfo,
                UseSkinSprites = Beatmap.Storyboard.UseSkinSprites,
            };
```

peppy replied:
That doesn't sound like it will scale.

peppy replied:
I think what you're going to want is to make a derived `DrawableStoryboard` which handles the filter operations internally. So you don't need to mutate the `Storyboard` in the first place.

LeNitrous replied:
I've done a bit more research on how the elements are created.

The elements in the `DrawableStoryboard` are deeply nested inside the `LayerManagementContainer`. The hierarchy looks like this:

```
DrawableStoryboard
├─ DrawableStoryboardLayer
│  ├─ LayerElementContainer
│  │  ├─ DrawableStoryboardSprite
```

In short, there is no easy access to the drawable storyboard elements in the first place without making major changes. I'm rather unsure if this is the intended direction or not.

peppy replied:
So to keep things simple then:

- There's no real reason to remove the fail layer right? It's not going to display either way? If so ignore it and just let it be.
- To mute sounds, you should be able to nest the whole `DrawableStoryboard` inside an `AudioContainer` and set the volume adjustment on that to 0.

LeNitrous committed:
apply suggestions

- apply 0 alpha to beatmap background if storyboard replaces it
- use an AudioContainer to mute all samples coming from the storyboard

peppy committed:
Tidy up code

peppy commented:
Returning to the main menu from song select causes the storyboard to fade then reload. Surely you noticed this while testing?

peppy commented:
> Returning to the main menu from song select causes the storyboard to fade then reload. Surely you noticed this while testing?

I'll look into this as a separate fix. Looks to be unrelated. I think we can just get this in and fix that separately.
